qiskit/quantum_info/states/stabilizerstate.py

```diff
@@ -0,0 +1,557 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Stabilizer state class.
+"""
+
+import copy
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.instruction import Instruction
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.states.quantum_state import QuantumState
+from qiskit.quantum_info.operators.symplectic import Clifford, Pauli
+from qiskit.quantum_info.operators.symplectic.clifford_circuits import _append_x
+
+
+class StabilizerState(QuantumState):
+    """StabilizerState class. Based on the internal class:
+    :class:`~qiskit.quantum_info.Clifford`
+
+    References:
+        1. S. Aaronson, D. Gottesman, *Improved Simulation of Stabilizer Circuits*,
+           Phys. Rev. A 70, 052328 (2004).
+           `arXiv:quant-ph/0406196 <https://arxiv.org/abs/quant-ph/0406196>`_
+    """
+
+    def __init__(self, data):
+        """Initialize a StabilizerState object.
+
+        Args:
+            data (StabilizerState or Clifford or Pauli or QuantumCircuit or
+                  qiskit.circuit.Instruction):
+                Data from which the stabilizer state can be constructed.
+
+        Raises:
+            QiskitError: if input data is not a valid stabilizer state (not a valid Clifford).
+        """
+
+        # Initialize from a Clifford
+        if isinstance(data, Clifford):
+            self._data = data
+
+        # Initialize from a Pauli
+        elif isinstance(data, Pauli):
+            self._data = Clifford(data.to_instruction())
+
+        # Initialize from another StabilizerState
+        elif isinstance(data, StabilizerState):
+            self._data = data._data
+
+        # Initialize from a QuantumCircuit or Instruction object
+        elif isinstance(data, (QuantumCircuit, Instruction)):
+            self._data = Clifford.from_circuit(data)
+
+        # Validate that Clifford table is a symplectic matrix
+        if not self.is_valid():
+            raise QiskitError(
+                'Invalid StabilizerState. Input is not a valid Clifford.')
+
+        # Initialize
+        super().__init__()
+
+    def __eq__(self, other):
+        return self._data == other._data
```

Comment:
This compares the tableau representations, but the same stabilizer state can have many tableau representations, so the usage of the `==` operator is misleading. Maybe it's better to avoid implementing `__eq__` at all, instead having two equality functions with descriptive names - one for comparing tableau and the other for comparing the state itself.

Reply:
The current `==` operator checks that both stabilizer tables are the same.
It's true that the same stabilizer state can be represented by several stabilizer tables, representing several sets of generators.
For example, the stabilizer group of a Bell state is `{II, XX, -YY, ZZ}`, hence it can be generated by `{XX, ZZ }`, `{XX, -YY}` or `{-YY, ZZ}`, which provide 3 different stabilizer tables.
Perhaps we should add an extra method called `equiv` that checks if two generating sets give rise to the same stabilizer group (and stabilizer state), but I don't know how to check it.

Reply:
I updated eq so it will check only the stabilizer part, but I don't know how to check equivalence of different stabilizer sets of the same stabilizer group.